### PR TITLE
Issue 170: upgrade extension for new version compatibility

### PR DESCRIPTION
**Pull request by Jawad-A02**
## Description

Please include a summary of the changes and the related issue. Please also include relevant motivation and context.

This should allow the extension to be compatible with the new schema and prompt format that agent mark v3 allows.
Couple of changes to allow this to work:
1. Change property model -> either `image_model` or `object_model` or `text_model` (bcz each model might have a different invoke method and properties needed)
2. Change property name -> `model_name`
3. Allow register model function to accept `provider` as an arg that will allow us to save and store the api keys for different providers for different models.
4. Change the extension implementation to rely on the `createAgent` class instead of runtime functions.
5. I have removed package-lock.json as the repo relies on yarn so it wasn't needed but lmk if I should add it again.

Fixes # [issue #170 ](https://github.com/puzzlet-ai/agentmark/issues/170)

## Type of Change

- [ ] Bug fix
- [x] New feature
- [ ] Documentation update
- [ ] Other (please specify)

## How Has This Been Tested?

1. **Ran the extension through 3 mdx files, each being a different config that agent mark allows.**
2. Tested the edge cases of providing multiple configs or no configs at all.
3. Also tested `agentMark `running by itself through the new changes.

Please describe the tests that you ran to verify your changes.

## Checklist:

- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: packages/agentmark/src/template_engines/templatedx.ts [hunk 1]
```diff
@@ -86,10 +86,12 @@ export async function getRawConfig(ast: Ast, props?: JSONObject) {
   const extractedFieldPromises = shared["__puzzlet-extractTextPromises"] || [];
   const messages = getMessages(await Promise.all(extractedFieldPromises));
 
-  frontMatter.model = frontMatter.model || {};
   return {
     name: frontMatter.name,
     messages: messages,
-    model: frontMatter.model,
+    image_model: frontMatter.image_model,
```
Comment (kocobee) [inline packages/agentmark/src/template_engines/templatedx.ts:92]: A few notes here:
1. The `test_settings` really don't belong in the AgentMark SDK, and are really only used for the extension (and our platform). These can be parsed separately by those.
2. We should use `_config` instead of `_model`. Sometimes the same models are used to generate 2 different things (i.e. text vs object).
3. I _think_ we should conditionally add these instead of attaching a bunch of undefined configs here, but I could be convinced otherwise.

File: packages/vscode-extension/src/extension.ts [hunk 2]
```diff
@@ -24,21 +41,28 @@ export function activate(context: vscode.ExtensionContext) {
...
       }
-      const ast = await load(file);
 
-      const model = getModel(ast);
-
-      const plugin = ModelPluginRegistry.getPlugin(model);
-
-      if (!plugin) {
-        return vscode.window.showErrorMessage(`Error: No Support for ${model}`);
+      const loadedFile = await loader.load(file);
```
Comment (kocobee) [inline packages/vscode-extension/src/extension.ts:45]: See slack notes about how to properly load an ast, and pass into agentmark

File: packages/vscode-extension/src/extension.ts [hunk 3]
```diff
@@ -24,21 +41,28 @@ export function activate(context: vscode.ExtensionContext) {
...
+      const yaml = await templateEngine.compile(loadedFile);
+      const modelEntries = [
```
Comment (kocobee) [inline packages/vscode-extension/src/extension.ts:47]: Nit: If statements are probably cleaner/more readable here. Not sure we need the mapping

File: packages/agentmark/src/adapters/vercel.ts [hunk 4]
```diff
@@ -88,20 +88,26 @@ export class VercelToolRegistry {
...
+  private providerMatches: Record<string, string> = {};
   private patternMatches: Array<[RegExp, ModelFunctionCreator]> = [];
   private defaultCreator?: ModelFunctionCreator;
 
   constructor(defaultCreator?: ModelFunctionCreator) {
     this.defaultCreator = defaultCreator;
   }
 
-  registerModel(modelPattern: string | RegExp | Array<string>, creator: ModelFunctionCreator): void {
+  registerModel(modelPattern: string | RegExp | Array<string>, creator: ModelFunctionCreator, provider: string): void {
```
Comment (kocobee) [inline packages/agentmark/src/adapters/vercel.ts:99]: Hmm, not sure the "provider" is required here. Instead, when you call `format` you can pass in `options`. This currently accepts an `apiKey`, so you can pass this in through there. The tests should have an example as well.

Comment (Jawad-A02) [reply packages/agentmark/src/adapters/vercel.ts:99]: So we used to store the key using the `plugin.provider` so now all the users that used `agentMark` before have their stored keys in the `context.secrets.store(`prompt-dx.${plugin.provider}`, apiKey);` The plugin provider was different for openAI and anthropic models so to ensure that the users didn't have to readd their keys, I added provider to the registerModel function.

Comment (kocobee) [reply packages/agentmark/src/adapters/vercel.ts:99]: I think this type of logic should be maintained in the extension vs AgentMark. I don't believe AM should know/care about providers. Instead, when you register via Vercel, you can just provide the api key (i.e. https://sdk.vercel.ai/providers/ai-sdk-providers/openai). So that logic can happen here: https://github.com/puzzlet-ai/agentmark/pull/174/files#diff-4c90b03b52f52f0b353937581a6f678abab846759e4ebe0fc8064e3004ff0164R6-R16

Comment (Jawad-A02) [reply packages/agentmark/src/adapters/vercel.ts:99]: Well it isn't really about providing the apiKey rather where to store the different apiKey for each provider? openAI keys will be different than anthro ones so this is just a simple way to add a string to storage key for the apiKey... I do see your point tho, what do you recommend I do to store the apiKey?

File: packages/vscode-extension/src/extension.ts (hunk 3 above)
Comment (kocobee) [reply packages/vscode-extension/src/extension.ts:35]: Just another reminder here

File: packages/vscode-extension/src/extension.ts [hunk 5]
```diff
@@ -1,16 +1,15 @@
...
-  getModel,
-  load,
-  streamInference,
-} from "@puzzlet/agentmark";
-import { getFrontMatter } from "@puzzlet/templatedx";
-import AllModelPlugins from '@puzzlet/all-models';
 import * as vscode from "vscode";
+import { createAgentMark, TemplateDXTemplateEngine, VercelAdapter } from "@puzzlet/agentmark";
+import { experimental_generateImage as generateImage, streamObject, streamText } from "ai";
+import { getFrontMatter, load } from "@puzzlet/templatedx";
```
Comment (kocobee) [inline packages/vscode-extension/src/extension.ts:4]: I don't remember if load should/shouldn't work just as-is. I would make sure you test some random mdx file locations to ensure it can run them all as expected, if you haven't already.

File: packages/agentmark/package.json [hunk 6]
```diff
@@ -21,22 +21,20 @@
...
     "@types/node": "^22.8.1",
     "@types/react": "^18",
-    "ai": "^4.0.0",
+    "cpy-cli": "^5.0.0",
     "react": "^18.3.1",
     "tsup": "^8.3.5",
     "typescript": "^5.6.2",
     "vitest": "^3.0.8"
   },
-  "peerDependencies": {
```
Comment (kocobee) [inline packages/agentmark/package.json:35]: Why did you remove "ai" as dev/peer deps?

File: package.json [hunk 7]
```diff
@@ -18,6 +18,7 @@
   "packageManager": "yarn@4.5.3",
   "devDependencies": {
     "@changesets/cli": "^2.27.9",
+    "ai": "^4.3.5",
```
Comment (kocobee) [inline package.json:21]: This shouldn't be in the root json

File: packages/agentmark/src/adapters/vercel.ts [hunk 8]
```diff
@@ -123,9 +129,21 @@ export class VercelModelRegistry {
     throw new Error(`No model function found for: ${modelName}`);
   }
 
-  registerModels(mappings: Record<string, ModelFunctionCreator>): void {
-    for (const [pattern, creator] of Object.entries(mappings)) {
-      this.registerModel(pattern, creator);
+  getProvider(modelName: string): string | undefined {
```
Comment (kocobee) [inline packages/agentmark/src/adapters/vercel.ts:132]: I don't think this logic belongs here. See alternative solution below.

File: packages/vscode-extension/src/extension.ts [hunk 9]
```diff
@@ -24,21 +23,35 @@ export function activate(context: vscode.ExtensionContext) {
...
+          ? "No config (image_config, object_config, or text_config) found in the file."
+          : "Only one config (image_config, object_config, or text_config) should be defined at a time.";
+        return vscode.window.showErrorMessage(message);
       }
 
-      let apiKey = await context.secrets.get(`prompt-dx.${plugin.provider}`);
+      const [modelConfig, model] = modelEntries[0];
+      const modelName = model?.model_name || '';
 
+      let apiKey = await context.secrets.get(`prompt-dx.${modelRegistry.getProvider(modelName)}`);
```
Comment (kocobee) [inline packages/vscode-extension/src/extension.ts:51]: I think we should maintain this mapping internally (in the extension) instead of in AgentMark. Alternatively, the AI SDK might provide this, but not sure. Or you can possibly do `model.constructor.name` (this might be easiest). Either way, let's remove the logic in AgentMark. We're trying to keep the API as minimal as possible.

File: packages/agentmark/src/agentmark.ts [hunk 10]
```diff
@@ -30,12 +29,12 @@ export class AgentMark<
...
 
   async loadTextPrompt<K extends keyof T & string>(
-    pathOrPreloaded: K, 
+    pathOrPreloaded: K,
     options?: any
   ): Promise<TextPrompt<T, A, K>> {
     let content: unknown;
     
-    if (typeof pathOrPreloaded === 'string') {
+    if (typeof pathOrPreloaded === 'string' && this.loader) {
```
[the remark this answers is not in the document]
Comment (Jawad-A02) [reply packages/agentmark/src/agentmark.ts:37]: Yup this is the idea cause loader would be undefined in that case

File: packages/vscode-extension/src/extension.ts [hunk 11]
```diff
@@ -24,21 +23,35 @@ export function activate(context: vscode.ExtensionContext) {
...
-      if (!plugin) {
-        return vscode.window.showErrorMessage(`Error: No Support for ${model}`);
+      const compiledYaml = await templateEngine.compile(ast);
+      const modelEntries = [
+        ["image_config", compiledYaml?.image_config],
+        ["object_config", compiledYaml?.object_config],
+        ["text_config", compiledYaml?.text_config]
+      ].filter(([_, val]) => Boolean(val)) as [modelConfig, any][];
+
+      if (modelEntries.length !== 1) {
```
Comment (kocobee) [inline packages/vscode-extension/src/extension.ts:41]: This error checking might not be needed. I believe this is already handled automatically when we load/check the schema(?)

File: packages/vscode-extension/src/extension.ts [hunk 12]
```diff
@@ -24,21 +23,51 @@ export function activate(context: vscode.ExtensionContext) {
...
+      let modelConfig: modelConfig | undefined;
+      let model: any;
+      
+      const definedConfigs = [
+        compiledYaml?.image_config,
+        compiledYaml?.object_config,
+        compiledYaml?.text_config,
+      ].filter(Boolean);
+      
+      if (definedConfigs.length > 1) {
```
Comment (kocobee) [inline packages/vscode-extension/src/extension.ts:45]: I don't think this check is necessary. The schema check when we load should handle this, and then we can delete the definedConfigs above as well.

File: packages/agentmark/test/agentmark.test.ts [hunk 13]
```diff
@@ -95,9 +95,9 @@ describe('AgentMark Integration', () => {
     expect(result.messages[2].role).toBe('assistant');
     expect(result.messages[2].content).toBe('Here\'s your answer!');
 
-    expect(result.model.name).toBe('test-model');
-    expect(result.model.schema).toBeDefined();
-    expect(result.model.schema.properties.answer).toBeDefined();
+    expect(result.object_config.model_name).toBe('test-model');
```
Comment (kocobee) [inline packages/agentmark/test/agentmark.test.ts:98]: Should we add a test somewhere for the image config as well?

Also, I'd like to see a test that checks only one config is defined (i.e. errors out if multiple are). This should ideally just work as-is w/ our schema check.

File: packages/agentmark/src/adapters/vercel.ts [hunk 14]
```diff
@@ -123,8 +125,8 @@ export class VercelModelRegistry {
     throw new Error(`No model function found for: ${modelName}`);
   }
 
-  registerModels(mappings: Record<string, ModelFunctionCreator>): void {
-    for (const [pattern, creator] of Object.entries(mappings)) {
+  registerModels(mappings: Record<string, { creator: ModelFunctionCreator}>): void {
```
Comment (kocobee) [inline packages/agentmark/src/adapters/vercel.ts:128]: I think we can probably axe this whole method and just rename the above to `registerModels` instead.

File: packages/agentmark/src/adapters/vercel.ts [hunk 15]
```diff
@@ -183,18 +185,22 @@ export class VercelAdapter<
   }
 
   adaptObject<K extends keyof T & string>(
-    input: ObjectConfig<Schema<T[K]["output"]>>,
+    input: Omit<ObjectConfig<Schema<T[K]["output"]>>, "object_model"> & {
```
Comment (kocobee) [inline packages/agentmark/src/adapters/vercel.ts:188]: Hmm. Not entirely sure what's happening here. Why was this changed?

File: packages/agentmark/test/agentmark.test.ts (hunk 13 above)
Comment (Jawad-A02) [reply packages/agentmark/test/agentmark.test.ts:98]: I wanted to add a test for images... but though it might be better to wait for the other issue?
https://github.com/agentmark-ai/agentmark/issues/168

File: packages/agentmark/src/adapters/vercel.ts (hunk 15 above)
Comment (Jawad-A02) [reply packages/agentmark/src/adapters/vercel.ts:188]: Be honest, I wasn't entirely sure, but for some reason changing the model to be `object_config`  made it error out cause `typedSchema` wasn't expected...

**Comment by kocobee** [comment]
Mind updating the README(s) w/ the corrected mdx examples?

File: packages/agentmark/src/prompts/object.ts [hunk 16]
```diff
@@ -23,15 +22,7 @@ export class ObjectPrompt<
     options: JSONObject = {}
   ): Promise<ReturnType<A['adaptObject']>> {
     const compiledTemplate = await this.templateEngine.compile(this.template, props) as ObjectConfig;
-    const typedSchema = jsonSchema<T[K]["output"]>(compiledTemplate.model.schema);
```
Comment (kocobee) [inline packages/agentmark/src/prompts/object.ts:26]: This needs to be here. We need to pass the `typedSchema`.

I'd expect something like: 

```
 const enhancedTemplate = {
      ...compiledTemplate,
      object_config: {
        ...compiledTemplate.object_config,
        typedSchema,
      },
    };
    ```

**Comment by kocobee** [comment]
Looks like the build is failing as well. Might need to run `yarn`

**Review by kocobee — APPROVED** [review]
Code LGTM. Will test more locally before V3 release.
